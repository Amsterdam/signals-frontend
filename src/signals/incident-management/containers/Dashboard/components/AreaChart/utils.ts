// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import { Direction } from './ComparisonRate'
import type { ComparisonRateType } from './ComparisonRate'
import type { AreaChartValue as Value, Today } from '../../charts/types'

export const formatData = (values: Value[]): Value[] => {
  return values.map((value, i) =>
    i === values.length - 1
      ? {
          ...value,
          date: `${value.date} 12:00`,
          image: '/assets/images/area-chart-tooltip.svg',
        }
      : { ...value, date: `${value.date} 12:00` }
  )
}

export const getToday = (): Today => {
  const today = new Date()

  return {
    year: today.getFullYear(),
    // Vega lite does not work with zero indexed months
    month: today.getMonth() + 1,
    date: today.getDate(),
    hours: 12,
    minutes: 0o0,
  }
}

export const getPercentage = (values: Value[]): ComparisonRateType => {
  const thisWeek = values.reduce((total, value) => total + value.amount, 0)
  const lastWeek = values.reduce(
    (total, value) => total + value.amount_week_earlier,
    0
  )

  return {
    direction: thisWeek < lastWeek ? Direction.DOWN : Direction.UP,
    percentage: Math.abs(Math.round((thisWeek / lastWeek - 1) * 100)),
  }
}
