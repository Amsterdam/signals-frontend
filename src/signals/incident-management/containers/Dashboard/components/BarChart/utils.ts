// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import configuration from 'shared/services/configuration/configuration'

import type { RawData } from './types'
import { constants } from '../../charts'
import type { BarChartValue } from '../../charts'

export const getQueryList = (queryString: string) => {
  const endpoint = configuration.INCIDENT_PRIVATE_ENDPOINT
  return constants.statusList.map((status) => {
    const params = queryString
      ? `${queryString}&status=${status.query}`
      : `status=${status.query}`
    return `${endpoint}stats/total?${params}`
  })
}

export const formatData = (rawData: RawData[]): BarChartValue[] => {
  return rawData.map((value, i) => {
    return {
      status: constants.statusList[i].label,
      nrOfIncidents: value.total,
    }
  })
}

export const getTotalNrOfIncidents = (rawData: RawData[]): number => {
  return rawData.reduce(
    (aggregatedTotal, value) => aggregatedTotal + value.total,
    0
  )
}
