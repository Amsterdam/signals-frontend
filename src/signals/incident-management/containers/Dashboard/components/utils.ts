// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type { IncidentCount } from '../charts/types'

export const getMaxDomain = (values: IncidentCount[]) =>
  Math.max(...values.map((v) => v.amount)) * 1.2
