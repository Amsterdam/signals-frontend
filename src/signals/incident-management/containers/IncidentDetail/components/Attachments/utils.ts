// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type { Attachment } from '../../types'

export const getAttachmentFileName = (location: Attachment['location']) =>
  location.split('?')[0].split('/').pop()
