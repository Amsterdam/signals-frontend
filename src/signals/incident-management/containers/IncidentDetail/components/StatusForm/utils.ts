// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2020 - 2021 Gemeente Amsterdam
import type { AlertLevel } from '@amsterdam/asc-ui'

import configuration from 'shared/services/configuration/configuration'
import {
  changeStatusOptionList,
  isStatusClosed,
} from 'signals/incident-management/definitions/statusList'
import { StatusCode } from 'types/status-code'

import * as constants from './constants'

export const emailSentWhenStatusChangedTo = ({
  toStatus,
  fromStatus,
  isSplitIncident,
}: {
  toStatus: StatusCode
  fromStatus: StatusCode
  isSplitIncident: boolean
}): boolean => {
  if (isSplitIncident) return false

  if (
    !configuration.featureFlags.reporterMailHandledNegativeContactEnabled &&
    fromStatus === StatusCode.VerzoekTotHeropenen &&
    toStatus === StatusCode.Afgehandeld
  ) {
    return false
  }

  return Boolean(
    changeStatusOptionList.find(
      ({ email_sent_when_set, key }) => email_sent_when_set && toStatus === key
    )
  )
}

export const textIsRequired = ({
  fromStatus,
  toStatus,
  isSplitIncident,
}: {
  fromStatus: StatusCode
  toStatus: StatusCode
  isSplitIncident: boolean
}): boolean => {
  if (
    isSplitIncident ||
    !configuration.featureFlags.reporterMailHandledNegativeContactEnabled
  ) {
    return [
      StatusCode.Afgehandeld,
      StatusCode.Ingepland,
      StatusCode.Heropend,
      StatusCode.ReactieGevraagd,
    ].includes(toStatus)
  } else {
    return emailSentWhenStatusChangedTo({
      fromStatus,
      toStatus,
      isSplitIncident,
    })
  }
}

type Warning = {
  key: string
  level: AlertLevel
  content: string
  heading?: string
}

export const determineWarnings = ({
  originalStatus,
  toStatus,
  isSplitIncident,
  hasEmail,
  hasOpenChildren,
  categorySlug,
}: {
  originalStatus: StatusCode
  toStatus: StatusCode
  isSplitIncident: boolean
  hasEmail: boolean
  hasOpenChildren: boolean
  categorySlug?: string
}): Warning[] => {
  const warnings: Warning[] = []

  if (isSplitIncident) return []

  if (isStatusClosed(toStatus) && hasOpenChildren) {
    warnings.push({
      key: 'has-open-child-incidents-warning',
      heading: constants.DEELMELDINGEN_STILL_OPEN_HEADING,
      content: constants.DEELMELDINGEN_STILL_OPEN_CONTENT,
      level: 'info',
    })
  }

  if (originalStatus === StatusCode.ReactieGevraagd && hasEmail) {
    warnings.push({
      key: 'has-open-reply-request-warning',
      heading: constants.REPLY_CHANGE_STATUS_HEADING,
      content: constants.REPLY_CHANGE_STATUS_CONTENT,
      level: 'info',
    })
  }

  if (
    configuration.featureFlags.disableClosingCategoryOverigOverig &&
    toStatus === StatusCode.Afgehandeld &&
    categorySlug === 'overig'
  ) {
    warnings.push({
      key: 'has-category-overig',
      heading: constants.CATEGORY_OVERIG_HEADING,
      content: constants.CATEGORY_OVERIG_CONTENT,
      level: 'error',
    })
  } else if (toStatus === StatusCode.Afgehandeld)
    warnings.push({
      key: 'end-status-warning',
      content: constants.AFGEHANDELD_CONTENT,
      level: 'neutral',
    })

  if (toStatus === StatusCode.ReactieGevraagd && !hasEmail) {
    warnings.push({
      key: 'has-no-email-reply-warning',
      heading: constants.REPLY_NO_MAIL_HEADING,
      content: constants.REPLY_NO_MAIL_CONTENT,
      level: 'error',
    })
  }

  return warnings
}

export const getTextConfig = (statusCode: StatusCode) => {
  return statusCode === StatusCode.ReactieGevraagd
    ? {
        label: constants.REPLY_MAIL_LABEL,
        maxLength: constants.REPLY_MAIL_MAX_LENGTH,
        rows: 6,
      }
    : {
        label: constants.DEFAULT_TEXT_LABEL,
        maxLength: constants.DEFAULT_TEXT_MAX_LENGTH,
        rows: 9,
      }
}
