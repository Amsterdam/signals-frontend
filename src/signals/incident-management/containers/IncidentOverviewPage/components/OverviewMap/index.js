import OverviewMap from './OverviewMap';

export default OverviewMap;
