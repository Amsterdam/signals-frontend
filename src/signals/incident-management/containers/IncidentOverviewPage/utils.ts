// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import { SortOptions } from './contants'

export function compareSortOptions(a: SortOptions, b: SortOptions) {
  return a.replace('-', '') === b.replace('-', '')
}

export const sortException = (sortOption: SortOptions) => {
  /** Priority is based on alphabetical order, therefore chevron should be reversed. */
  return (
    sortOption === SortOptions.PRIORITY_ASC ||
    sortOption === SortOptions.PRIORITY_DESC
  )
}
