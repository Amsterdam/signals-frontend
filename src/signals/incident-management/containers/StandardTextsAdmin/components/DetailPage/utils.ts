// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type { StandardTextDetailData, StandardTextForm } from './types'

export const createPatch = (
  data: StandardTextForm,
  dirtyFields: Partial<
    Readonly<{
      categories?: boolean[] | undefined
      state?: boolean | undefined
      title?: boolean | undefined
      text?: boolean | undefined
      active?: boolean | undefined
    }>
  >
): StandardTextDetailData => {
  /* istanbul ignore next */
  const payload = Object.keys(dirtyFields).map((key) => {
    return {
      [key]: data[key as keyof StandardTextForm],
    }
  })

  return Object.assign({}, ...payload)
}

export const createPost = (data: StandardTextForm): StandardTextDetailData => {
  const payload = Object.entries(data).map(([key, value]) => {
    return { [key]: value }
  })
  return Object.assign({}, ...payload)
}
