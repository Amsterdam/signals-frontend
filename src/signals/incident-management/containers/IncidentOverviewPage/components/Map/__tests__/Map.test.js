import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import geographyJSON from 'utils/__tests__/fixtures/geography.json';
import { withMapContext } from 'test/utils';
import OverviewMap from '..';

fetch.mockResponse(JSON.stringify(geographyJSON));

describe('signals/incident-management/containers/IncidentOverviewPage/components/Map', () => {
  afterEach(() => {
    fetch.resetMocks();
  });

  it('should render the map and the autosuggest', async () => {
    const { getByTestId, findByTestId } = render(withMapContext(<OverviewMap />));

    await findByTestId('map-base');

    expect(getByTestId('autoSuggest')).toBeInTheDocument();
  });

  it('should request locations', async () => {
    const { findByTestId, rerender, unmount } = render(withMapContext(<OverviewMap />));

    await findByTestId('map-base');

    const reDateTime = /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/;
    const expectedFilterParams = {
      created_after: reDateTime,
      created_before: reDateTime,
      page_size: /4000/,
    };

    expect(fetch.mock.calls).toHaveLength(1);

    const requestUrl = new URL(fetch.mock.calls[0][0]);
    const params = new URLSearchParams(requestUrl.search);

    Object.keys(expectedFilterParams).forEach(key => {
      expect([...params.keys()].includes(key)).toBeTruthy();

      const [, param] = [...params.entries()].find(([k]) => k === key);
      expect(param).toMatch(expectedFilterParams[key]);
    });

    unmount();

    rerender(withMapContext(<OverviewMap />));

    await findByTestId('map-base');

    expect(fetch.mock.calls).toHaveLength(2);
  });

  it('should render detail panel', async () => {
    const { getByTestId, findByTestId } = render(withMapContext(<OverviewMap showPanelOnInit />));

    await findByTestId('map-base');

    expect(getByTestId('mapDetailPanel')).toBeInTheDocument();
  });

  it('should close detail panel', async () => {
    const { queryByTestId, findByTestId } = render(withMapContext(<OverviewMap showPanelOnInit />));

    const detailPanel = await findByTestId('mapDetailPanel');

    act(() => {
      fireEvent.click(detailPanel.querySelector('button'));
    });

    expect(queryByTestId('mapDetailPanel')).not.toBeInTheDocument();
  });
});
