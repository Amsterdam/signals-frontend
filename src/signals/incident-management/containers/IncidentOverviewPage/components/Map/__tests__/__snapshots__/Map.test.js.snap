// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signals/incident-management/containers/IncidentOverviewPage/components/Map should render detail panel 1`] = `
<div
  class="Map__Wrapper-q7dcmt-0 jZrEsF"
>
  <div
    class="Map__StyledMap-q7dcmt-1 dQygRr leaflet-container leaflet-grab leaflet-touch-drag"
    data-testid="overviewMap"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t4.data.amsterdam.nl/topo_rd/10/472/540.png"
              style="width: 256px; height: 256px; left: 282px; top: 339px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t3.data.amsterdam.nl/topo_rd/10/472/541.png"
              style="width: 256px; height: 256px; left: 282px; top: 83px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t3.data.amsterdam.nl/topo_rd/10/471/540.png"
              style="width: 256px; height: 256px; left: 26px; top: 339px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t1.data.amsterdam.nl/topo_rd/10/473/540.png"
              style="width: 256px; height: 256px; left: 538px; top: 339px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t1.data.amsterdam.nl/topo_rd/10/472/539.png"
              style="width: 256px; height: 256px; left: 282px; top: 595px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t4.data.amsterdam.nl/topo_rd/10/473/541.png"
              style="width: 256px; height: 256px; left: 538px; top: 83px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t2.data.amsterdam.nl/topo_rd/10/471/541.png"
              style="width: 256px; height: 256px; left: 26px; top: 83px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t4.data.amsterdam.nl/topo_rd/10/471/539.png"
              style="width: 256px; height: 256px; left: 26px; top: 595px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t2.data.amsterdam.nl/topo_rd/10/473/539.png"
              style="width: 256px; height: 256px; left: 538px; top: 595px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t2.data.amsterdam.nl/topo_rd/10/470/540.png"
              style="width: 256px; height: 256px; left: -230px; top: 339px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t2.data.amsterdam.nl/topo_rd/10/472/538.png"
              style="width: 256px; height: 256px; left: 282px; top: 851px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t2.data.amsterdam.nl/topo_rd/10/474/540.png"
              style="width: 256px; height: 256px; left: 794px; top: 339px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t2.data.amsterdam.nl/topo_rd/10/472/542.png"
              style="width: 256px; height: 256px; left: 282px; top: -173px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t3.data.amsterdam.nl/topo_rd/10/470/539.png"
              style="width: 256px; height: 256px; left: -230px; top: 595px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t3.data.amsterdam.nl/topo_rd/10/473/542.png"
              style="width: 256px; height: 256px; left: 538px; top: -173px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t1.data.amsterdam.nl/topo_rd/10/470/541.png"
              style="width: 256px; height: 256px; left: -230px; top: 83px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t1.data.amsterdam.nl/topo_rd/10/474/541.png"
              style="width: 256px; height: 256px; left: 794px; top: 83px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t3.data.amsterdam.nl/topo_rd/10/474/539.png"
              style="width: 256px; height: 256px; left: 794px; top: 595px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t1.data.amsterdam.nl/topo_rd/10/471/538.png"
              style="width: 256px; height: 256px; left: 26px; top: 851px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t3.data.amsterdam.nl/topo_rd/10/473/538.png"
              style="width: 256px; height: 256px; left: 538px; top: 851px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t1.data.amsterdam.nl/topo_rd/10/471/542.png"
              style="width: 256px; height: 256px; left: 26px; top: -173px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t4.data.amsterdam.nl/topo_rd/10/474/542.png"
              style="width: 256px; height: 256px; left: 794px; top: -173px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t4.data.amsterdam.nl/topo_rd/10/470/542.png"
              style="width: 256px; height: 256px; left: -230px; top: -173px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t4.data.amsterdam.nl/topo_rd/10/470/538.png"
              style="width: 256px; height: 256px; left: -230px; top: 851px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://t4.data.amsterdam.nl/topo_rd/10/474/538.png"
              style="width: 256px; height: 256px; left: 794px; top: 851px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-shadow-pane leaflet-zoom-hide"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      >
        <canvas
          height="1229"
          style="left: -102px; top: -102px; width: 1229px; height: 1229px;"
          width="1229"
        />
      </div>
      <div
        class="leaflet-pane leaflet-marker-pane leaflet-zoom-hide"
      >
        <img
          alt=""
          class="leaflet-marker-icon map-marker-incident leaflet-zoom-hide leaflet-interactive"
          src="ImageMock"
          style="margin-left: -20px; margin-top: -39px; width: 40px; height: 40px; left: 795px; top: 1087px; z-index: 1087;"
          tabindex="0"
        />
        <div
          class="leaflet-marker-icon marker-cluster marker-cluster-small leaflet-zoom-hide leaflet-interactive"
          style="margin-left: -20px; margin-top: -20px; width: 40px; height: 40px; left: 358px; top: 609px; z-index: 609;"
          tabindex="0"
        >
          <div>
            <span>
              2
            </span>
          </div>
        </div>
        <div
          class="leaflet-marker-icon marker-cluster marker-cluster-small leaflet-zoom-hide leaflet-interactive"
          style="margin-left: -20px; margin-top: -20px; width: 40px; height: 40px; left: -405px; top: 228px; z-index: 228;"
          tabindex="0"
        >
          <div>
            <span>
              2
            </span>
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      />
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
           | Kaartgegevens CC-BY-4.0 Gemeente Amsterdam
        </div>
      </div>
    </div>
  </div>
  <div
    class="sc-cmthru bEhulO"
  >
    <div
      class="sc-hMFtBS gspQyK"
    >
      <div
        class="AutoSuggest__Wrapper-sc-1l4rsm2-0 eHfHIv Map__Autosuggest-q7dcmt-2 dPyNBw"
        data-testid="autoSuggest"
      >
        <div
          aria-controls="as-listbox"
          aria-expanded="false"
          aria-haspopup="listbox"
          role="combobox"
        >
          <div
            class="Input__Wrapper-sc-1933jsg-4 hZBxPn"
          >
            
            
            
            <input
              aria-autocomplete="list"
              class="Input__StyledInput-sc-1933jsg-1 cIXvqe"
              id=""
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-hMFtBS jhkBcj"
    />
    <div
      class="sc-hMFtBS fEpxzh"
    />
    <div
      class="sc-hMFtBS hVaEyk"
    />
  </div>
</div>
`;
